### PR TITLE
feat: support EIP-5792 methods

### PR DESCRIPTION
**Pull request by matthewwalsh0**
Support the EIP-5792 methods:

- `wallet_sendCalls`
- `wallet_getCallsStatus`
- `wallet_getCapabilities`

Specifically:

- Modularise new method support into separate files under `src/methods`.
  - As opposed to continually extending `src/wallet.ts`.
- Use `@metamask/superstruct` to easily validate request parameters against a  schema / struct.
- Move common validation functions to new `src/validation.ts`.
- Export new types inferred from structs.
  - Specifically as opposed to `*` for easier API diffs.

An additional benefit of using `@metamask/superstruct` is including multiple validation errors within a single RPC error:

```
Error: Invalid params

0 > from - Expected a string matching \`/^0x[0-9a-fA-F]{40}$/\` but received "123"
0 > chainId - Expected a string, but received: 123
0 > calls - Expected an array value, but received: "123"
```

### CODE REVIEW

File: src/methods/wallet-get-calls-status.ts [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+const GetCallsStatusReceiptStruct = object({
+  logs: array(
+    object({
+      address: HexChecksumAddressStruct,
+      data: StrictHexStruct,
+      topics: array(StrictHexStruct),
+    }),
+  ),
+  status: StrictHexStruct,
+  chainId: optional(StrictHexStruct),
```
Comment (jiexi) [inline src/methods/wallet-get-calls-status.ts:32]: is chainId optional in the result? I know we mean for it to be optional in the params of these new methods

File: src/methods/wallet-get-calls-status.ts [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+    getTransactionReceiptsByBatchId?: GetTransactionReceiptsByBatchIdHook;
+  },
+): Promise<void> {
+  if (!getTransactionReceiptsByBatchId) {
+    throw rpcErrors.methodNotSupported();
+  }
+
+  if (!validateParams(req.params, GetCallsStatusStruct)) {
+    return;
+  }
```
Comment (jiexi) [inline src/methods/wallet-get-calls-status.ts:67]: should this throw an error instead?

File: src/methods/wallet-get-calls-status.ts [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+export type GetCallsStatusReceipt = Infer<typeof GetCallsStatusReceiptStruct>;
+
+export type GetCallsStatusResult = {
+  status: 'PENDING' | 'CONFIRMED';
+  receipts?: GetCallsStatusReceipt[];
+};
+
+export type GetTransactionReceiptsByBatchIdHook = (
+  batchId: string,
+  req: JsonRpcRequest,
```
Comment (jiexi) [inline src/methods/wallet-get-calls-status.ts:49]: just curious why the whole req object is needed in the params since it isn't obvious based on the hook name

File: src/methods/wallet-get-calls-status.test.ts [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+    ]);
+  });
+
+  it('returns null if no receipts', async () => {
+    getTransactionReceiptsByBatchIdMock = jest.fn().mockResolvedValue([]);
+
+    await callMethod();
+    expect(response.result).toBeNull();
+  });
+
```
Comment (jiexi) [inline src/methods/wallet-get-calls-status.test.ts:94]: is it necessary to test the `mask()` call?

File: src/methods/wallet-get-calls-status.test.ts [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+      request,
+    );
+  });
+
+  it('returns confirmed status if all receipts available', async () => {
+    await callMethod();
+    expect(response.result?.status).toBe('CONFIRMED');
+  });
+
+  it('returns pending status if missing receipts', async () => {
```
Comment (jiexi) [inline src/methods/wallet-get-calls-status.test.ts:70]: should we also test that `receipts` `null` if there are pending?

File: src/methods/wallet-get-capabilities.ts [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+  PendingJsonRpcResponse,
+} from '@metamask/utils';
+import { HexChecksumAddressStruct } from '@metamask/utils';
+
+import { validateParams } from '../utils/validation';
+
+const GetCapabilitiesStruct = tuple([HexChecksumAddressStruct]);
+
+export type GetCapabilitiesParams = Infer<typeof GetCapabilitiesStruct>;
+export type GetCapabilitiesResult = Record<Hex, Record<string, any>>;
```
Comment (jiexi) [inline src/methods/wallet-get-capabilities.ts:17]: i know the specs say this is `any`, but is it actually `any`?

File: src/methods/wallet-get-capabilities.ts [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+  }: {
+    getCapabilities?: GetCapabilitiesHook;
+  },
+): Promise<void> {
+  if (!getCapabilities) {
+    throw rpcErrors.methodNotSupported();
+  }
+
+  if (!validateParams(req.params, GetCapabilitiesStruct)) {
+    return;
```
Comment (jiexi) [inline src/methods/wallet-get-capabilities.ts:38]: should we throw here?

File: src/methods/wallet-send-calls.ts [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+  res: PendingJsonRpcResponse<Json>,
+  {
+    getAccounts,
+    processSendCalls,
+  }: {
+    getAccounts: (req: JsonRpcRequest) => Promise<string[]>;
+    processSendCalls?: ProcessSendCallsHook;
+  },
+): Promise<void> {
+  if (!processSendCalls) {
```
Comment (jiexi) [inline src/methods/wallet-send-calls.ts:59]: should this throw is getAccounts is not provided as well?

File: src/methods/wallet-send-calls.ts [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+import { HexChecksumAddressStruct, StrictHexStruct } from '@metamask/utils';
+
+import {
+  validateAndNormalizeKeyholder,
+  validateParams,
+} from '../utils/validation';
+
+const SendCallsStruct = tuple([
+  object({
+    version: optional(nonempty(string())),
```
Comment (jiexi) [inline src/methods/wallet-send-calls.ts:26]: this doesn't seem to be optional in the spec

File: src/methods/wallet-send-calls.ts [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+    getAccounts: (req: JsonRpcRequest) => Promise<string[]>;
+    processSendCalls?: ProcessSendCallsHook;
+  },
+): Promise<void> {
+  if (!processSendCalls) {
+    throw rpcErrors.methodNotSupported();
+  }
+
+  if (!validateParams(req.params, SendCallsStruct)) {
+    // Not possible as throws.
```
Comment (jiexi) [inline src/methods/wallet-send-calls.ts:64]: ohhh i see. if `validateParams` throws, why do we need it in an if statement in each of these handlers?

File: src/methods/wallet-send-calls.test.ts [hunk 11]
```diff
@@ -0,0 +1,176 @@
...
+    params[0].version = '1.0.0';
+    await callMethod();
+    expect(processSendCallsMock).toHaveBeenCalledWith(params[0], request);
+  });
+
+  it('supports capabilities', async () => {
+    params[0].capabilities = { test: 'value' };
+    await callMethod();
+    expect(processSendCallsMock).toHaveBeenCalledWith(params[0], request);
+  });
```
Comment (jiexi) [inline src/methods/wallet-send-calls.test.ts:78]: these seem a bit redundant to me given what `calls hook` tests

File: src/methods/wallet-send-calls.test.ts [hunk 12]
```diff
@@ -0,0 +1,176 @@
...
+    await expect(callMethod()).rejects.toMatchInlineSnapshot(`
+            [Error: Invalid params
+
+            0 > from - Expected a string matching \`/^0x[0-9a-fA-F]{40}$/\` but received "0x123"
+            0 > calls > 0 > to - Expected a string matching \`/^0x[0-9a-fA-F]{40}$/\` but received "0x123"]
+          `);
+  });
+
+  it('throws if from is not in accounts', async () => {
+    getAccountsMock.mockResolvedValueOnce([]);
```
Comment (jiexi) [inline src/methods/wallet-send-calls.test.ts:170]: should `validateAndNormalizeKeyholder` be mocked instead?

File: src/utils/validation.ts [hunk 13]
```diff
@@ -0,0 +1,65 @@
...
+export async function validateAndNormalizeKeyholder(
+  address: Hex,
+  req: JsonRpcRequest,
+  { getAccounts }: { getAccounts: (req: JsonRpcRequest) => Promise<string[]> },
+): Promise<Hex> {
+  if (
+    typeof address === 'string' &&
+    address.length > 0 &&
+    resemblesAddress(address)
+  ) {
```
Comment (jiexi) [inline src/utils/validation.ts:15]: is there some helper in utils we could use here?

File: src/utils/validation.ts [hunk 14]
```diff
@@ -0,0 +1,65 @@
...
+
+  throw rpcErrors.invalidParams({
+    message: `Invalid parameters: must provide an Ethereum address.`,
+  });
+}
+
+export function validateParams<ParamsType>(
+  value: unknown | ParamsType,
+  struct: Struct<ParamsType>,
+): value is ParamsType {
```
Comment (jiexi) [inline src/utils/validation.ts:41]: seems like you might want the following return type instead:

```suggestion
): asserts value is ParamsType {
```

This would allow you to drop the return value in this helper and also the if statement guards in the handlers that call this

File: src/utils/validation.ts (hunk 13 above)
Comment (jiexi) [reply src/utils/validation.ts:15]: ahh i see. you're just DRYing this helper up. ehh okay, let's just leave as is then

File: src/methods/wallet-get-calls-status.ts (hunk 1 above)
Comment (matthewwalsh0) [reply src/methods/wallet-get-calls-status.ts:34]: This struct is only a means to remove any excess properties.

We don't want to validate the wallet itself so I've made all the properties optional, though I'll follow up to confirm why `chainId` was missing in my tests.

File: src/methods/wallet-get-calls-status.ts (hunk 3 above)
Comment (matthewwalsh0) [reply src/methods/wallet-get-calls-status.ts:51]: This is standard in the other hooks, as a way to provide any properties set by other middleware such as `networkClientId`.

File: src/methods/wallet-get-calls-status.test.ts (hunk 4 above)
Comment (matthewwalsh0) [reply src/methods/wallet-get-calls-status.test.ts:95]: Done.

File: src/methods/wallet-get-calls-status.test.ts (hunk 5 above)
Comment (matthewwalsh0) [reply src/methods/wallet-get-calls-status.test.ts:70]: Done.

File: src/methods/wallet-get-capabilities.ts (hunk 6 above)
Comment (matthewwalsh0) [reply src/methods/wallet-get-capabilities.ts]: Fair point, have changed to `Json` since it's being sent over JSON-RPC so is technically limited.

File: src/methods/wallet-get-capabilities.ts (hunk 7 above)
Comment (matthewwalsh0) [reply src/methods/wallet-get-capabilities.ts]: Removed in favour of `asserts`.

File: src/methods/wallet-send-calls.ts (hunk 8 above)
Comment (matthewwalsh0) [reply src/methods/wallet-send-calls.ts:59]: This is validated when `createWalletMiddleware` is called.

File: src/methods/wallet-send-calls.ts [hunk 15]
```diff
@@ -0,0 +1,80 @@
...
+
+export async function walletSendCalls(
+  req: JsonRpcRequest,
+  res: PendingJsonRpcResponse<Json>,
+  {
+    getAccounts,
+    processSendCalls,
+  }: {
+    getAccounts: (req: JsonRpcRequest) => Promise<string[]>;
+    processSendCalls?: ProcessSendCallsHook;
```
[the remark this answers is not in the document]
Comment (matthewwalsh0) [reply src/methods/wallet-send-calls.ts:56]: Yes, as otherwise we conceptually shouldn't need to validate it is provided below.

File: src/methods/wallet-send-calls.ts (hunk 9 above)
Comment (matthewwalsh0) [reply src/methods/wallet-send-calls.ts]: It's not technically required yet as we could assume no version is equivalent to `1.0`.

But I've made it required to be explicit and compliant with the spec.

File: src/methods/wallet-send-calls.ts (hunk 10 above)
Comment (matthewwalsh0) [reply src/methods/wallet-send-calls.ts]: To enforce the type guard, but no longer required with `asserts`.

File: src/methods/wallet-send-calls.test.ts (hunk 12 above)
Comment (matthewwalsh0) [reply src/methods/wallet-send-calls.test.ts:165]: A trade-off between test overlap and fragility. Since it was a one liner, I opted to treat the util as an implementation detail and not couple the unit tests to it.

File: src/utils/validation.ts (hunk 13 above)
Comment (matthewwalsh0) [reply src/utils/validation.ts:15]: Exactly, I wanted to use `isValidHexAddress` but didn't want to impact all the existing usages.

File: src/utils/validation.ts (hunk 14 above)
Comment (matthewwalsh0) [reply src/utils/validation.ts]: Done.

**Review by OGPoyraz — APPROVED** [review]

File: src/methods/wallet-send-calls.test.ts (hunk 12 above)
Comment (jiexi) [reply src/methods/wallet-send-calls.test.ts:165]: Fair point. I struggle with striking a balance in this area. Cool thanks!

**Review by jiexi — APPROVED** [review]

File: src/utils/validation.test.ts [hunk 16]
```diff
@@ -0,0 +1,128 @@
...
+      expect(resemblesAddress(ADDRESS_MOCK)).toBe(true);
+    });
+
+    it('returns false if length not correct', () => {
+      expect(resemblesAddress('0x123')).toBe(false);
+    });
+  });
+
+  describe('validateParams', () => {
+    it('does now throw if superstruct returns no error', () => {
```
Comment (cryptodev-2s) [inline src/utils/validation.test.ts:111]: Nit:
Small typo not instead of now

**Review by cryptodev-2s — APPROVED** [review]
LGTM, I left a tiny typo fix comment

File: src/utils/validation.test.ts (hunk 16 above)
Comment (matthewwalsh0) [reply src/utils/validation.test.ts:111]: Will get in any subsequent PR, thanks!
